elements/pl-python-variable/pl-python-variable.py

```diff
@@ -1,52 +1,119 @@
         var_out = pl.from_json(var_out)
 
+    html_list: List[str] = []
+
     # render the output variable
-    if var_type == 'dataframe':
-        html += var_out.to_html(header=show_header, index=show_index, classes=['pl-python-variable-table'])
+    if isinstance(var_out, pandas.DataFrame) and not force_text:
+        frame = cast(pandas.DataFrame, var_out)
+        frame_style = frame.style
+
+        if show_dtype:
+            descriptors = frame.agg([lambda s: s.dtype])
+            descriptors.index = ["dtype"]
+            other = descriptors.style.applymap(lambda v: "font-weight: bold;")
+            frame_style.set_table_styles(
+                [{"selector": ".foot_row0", "props": "border-top: 1px solid black;"}]
+            )
+            frame_style.concat(other)
+
+        if not show_header:
+            frame_style.hide(axis="columns")
+
+        if not show_index:
+            frame_style.hide()
+
+        # Might be worth moving everything out of the CSS file and handle it all with the builtin styler.
+        frame_style.set_table_attributes("class=pl-python-variable-table")
+
+        html_list.append(frame_style.to_html())
         if show_dimensions:
-            html += '<p class="pl-python-variable-table-dimensions">{} rows x {} columns</p>'.format(str(var_out.shape[0]), str(var_out.shape[1]))
+            html_list.append(
+                f'<p class="pl-python-variable-table-dimensions">{frame.shape[0]} rows x {frame.shape[1]} columns</p>'
+            )
     else:
-        no_highlight = pl.get_boolean_attrib(element, 'no-highlight', NO_HIGHLIGHT_DEFAULT)
-        prefix = pl.get_string_attrib(element, 'prefix', PREFIX_DEFAULT)
-        suffix = pl.get_string_attrib(element, 'suffix', SUFFIX_DEFAULT)
+        no_highlight = pl.get_boolean_attrib(
+            element, "no-highlight", NO_HIGHLIGHT_DEFAULT
+        )
+        prefix = pl.get_string_attrib(element, "prefix", PREFIX_DEFAULT)
+        suffix = pl.get_string_attrib(element, "suffix", SUFFIX_DEFAULT)
+
+        var_string = get_var_string(var_out, add_line_breaks)
+
+        html_list.append(
+            f'<pl-code language="python" no-highlight="{no_highlight}">{prefix}{var_string}{suffix}</pl-code>'
+        )
+
+    return "".join(html_list)
+
+
+def get_var_string(var: Any, add_line_breaks: bool) -> str:
+    if add_line_breaks:
+        if isinstance(var, dict):
+            inner_string = ",\n".join(
+                f"    {repr(key)}: {repr(val)}" for key, val in var.items()
+            )
+            return f"{{\n{inner_string}\n}}"
+
+        elif isinstance(var, list):
+            inner_string = ",\n".join(f"    {repr(elem)}" for elem in var)
+            return f"[\n{inner_string}\n]"
 
-        text = prefix + repr(var_out) + suffix
-        html += '<pl-code language="python" no-highlight="{}">{}</pl-code>'.format(no_highlight, text)
+        raise ValueError(f"Line breaks not supported for type {type(var).__name__}")
```

Comment:
Should this be promoted to `python/prairielearn.py`?

Reply:
I'm not so sure. I actually don't know of many places where we have to produce a nicer string corresponding to an internal collection (i.e. set or dict).

Reply:
Unless this code is used in at least one other element, it should not be moved there.